// sonda/dist/entrypoints/angular.cjs
"use strict";
const require_src = require('../src.cjs');
const require_esbuild = require('../esbuild.cjs');
const path = require_src.__toESM(require("path"));
const fs = require_src.__toESM(require("fs"));

//#region src/entrypoints/angular.ts
function SondaAngular(options = {}) {
	const cwd = process.cwd();
	const { config = "angular.json", projects = [],...opts } = options;
	opts.format ??= "html";
	opts.filename ??= `sonda-report-[project].${opts.format}`;
	if (!opts.filename.includes("[project]")) throw new Error("SondaAngular: The \"filename\" option must include the \"[project]\" token.");
	const angularConfig = loadJson(config);
	const projectsToGenerate = projects.length ? projects : Object.keys(angularConfig.projects);
	for (const project of projectsToGenerate) {
		const { outputPath } = angularConfig.projects[project].architect.build.options;
		const paths = typeof outputPath === "object" ? outputPath : { base: outputPath };
		paths.base = (0, path.resolve)(cwd, paths.base);
		paths.browser = (0, path.resolve)(paths.base, paths.browser || "browser");
		paths.server = (0, path.resolve)(paths.base, paths.server || "server");
		const metafile = updateMetafile(loadJson((0, path.resolve)(paths.base, "stats.json")), paths);
		const sondaOptions = Object.assign({}, opts);
		sondaOptions.filename = sondaOptions.filename.replace("[project]", project);
		require_esbuild.processEsbuildMetaFile(metafile, sondaOptions);
	}
}
function loadJson(path$1) {
	return JSON.parse((0, fs.readFileSync)((0, path.resolve)(process.cwd(), path$1), "utf8"));
}
/**
* Output paths in metafile only include file name, without the relative path from the current
* working directory. For example, in the metafile the output path is "main-xxx.js", but in the
* file system it's "dist/project/browser/en/main-xxx.js". This function updates the output paths
* to include the relative path from the current working directory.
*/
function updateMetafile(metafile, paths) {
	const cwd = process.cwd();
	const outputs = Object.assign({}, metafile.outputs);
	metafile.outputs = {};
	for (const path$1 of (0, fs.readdirSync)(paths.base, {
		encoding: "utf8",
		recursive: true
	})) {
		const absolutePath = (0, path.resolve)(paths.base, path$1);
		const filename = (0, path.basename)(absolutePath);
		const originalOutput = outputs[filename];
		if (originalOutput) metafile.outputs[(0, path.relative)(cwd, absolutePath)] = originalOutput;
	}
	return metafile;
}

//#endregion
module.exports = SondaAngular;
